Sync/BatchingClient.swift

```diff
@@ -0,0 +1,245 @@
+        return [NSURLQueryItem(name: "batch", value: String(token))]
+    }
+
+    private var bytes: ByteCount = 0
+    private var count: Int = 0
+    private var token: BatchToken
+
+    init(config: InfoConfiguration, token: BatchToken) {
+        self.config = config
+        self.token = token
+    }
+
+    mutating func add(payload: Payload) {
+        count += payload.count
+        bytes += payload.bytes
+    }
+}
+
+
+// MARK: Batching Client
+
+typealias BatchUploadFunction = (lines: [String], ifUnmodifiedSince: Timestamp?, queryParams: [NSURLQueryItem]?) -> Deferred<Maybe<StorageResponse<POSTResult>>>
+
+public class Sync15BatchClient<T: CleartextPayloadJSON> {
+    private(set) var ifUnmodifiedSince: Timestamp?
+
+    private let config: InfoConfiguration
+    private let uploader: BatchUploadFunction
+    private let serializeRecord: (Record<T>) -> String?
+
+    private var currentBatch: Batch?
+    private var currentPayload: Payload
+
+    private var onCollectionUploaded: (POSTResult, Timestamp?) -> DeferredTimestamp
+
+    init(config: InfoConfiguration, ifUnmodifiedSince: Timestamp? = nil, serializeRecord: (Record<T>) -> String?,
+         uploader: BatchUploadFunction, onCollectionUploaded: (POSTResult, Timestamp?) -> DeferredTimestamp) {
+        self.config = config
+        self.ifUnmodifiedSince = ifUnmodifiedSince
+        self.uploader = uploader
+        self.serializeRecord = serializeRecord
+
+        self.currentPayload = Payload(config: config)
+        self.onCollectionUploaded = onCollectionUploaded
+    }
+
+    public func endBatch() -> Success {
+        func uploadPayload() -> Success {
+            let lines = self.currentPayload.records.map { $0.payload }
+            let batchStartParam = NSURLQueryItem(name: "batch", value: "true")
+            let commitParam = NSURLQueryItem(name: "commit", value: "true")
+            return self.uploader(lines: lines, ifUnmodifiedSince: self.ifUnmodifiedSince, queryParams: [batchStartParam, commitParam])
+                >>== effect(moveForward) >>> succeed
+        }
+
+        // Send up as a single upload if we're not batching.
+        guard let batch = self.currentBatch else {
+            return self.currentPayload.isEmpty ? succeed() : uploadPayload()
+        }
+
+        // If we have a batch and both it and the payload are empty, don't need to do anything
+        if batch.isEmpty && self.currentPayload.isEmpty {
+            return succeed()
+        }
+
+        if batch.isFull {
+            return commit(batch) >>> uploadPayload
+        } else {
+            return commit(batch) >>> succeed
+        }
+    }
+
+    public func addRecords(records: [Record<T>]) -> Success {
+        if records.isEmpty {
+            return succeed()
+        }
+
+        do {
+            var generator = records.generate()
+            while let record = generator.next() {
+                let serializedRecord = try serialize(record)
```

Comment:
It's important to aggressively serialize outgoing records.

If you take a look at `IndependentRecordSerializer`, you'll see that if any record we plan to upload is over-sized, we upload *nothing*.

That's not the case with your refactoring here, if the server doesn't support batching (which will be the case for a while): we will accumulate the first fifty records and start an upload *without trying to serialize the 51st*. If a later record is too large, we will give up, leaving a partial write on the server.

So rework `addRecords` to be like this:

```
public func addRecords(records: [Record<T>]) -> Success {
    if records.isEmpty {
        return succeed()
    }
    do {
        let serialized = try records.map { try serialize($0) }     // or however Swift likes it
        return addRecords(serialized.generate())
    } catch let e {
        return deferMaybe(e as! MaybeErrorType)
    }
}

public func addRecords(records: Generator<UploadRecord>) -> Success {
    …
}
```